### PR TITLE
`General`: Request data export for another user as administrator

### PR DESCRIPTION
**Pull request by tobias-lippert**
#### General

- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.cit.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/development-process/#naming-conventions-for-github-pull-requests).
#### Server
- [x] **Important**: I implemented the changes with a very good performance and prevented too many (unnecessary) database calls.
- [x] I followed the [coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/).
- [x] I added multiple integration tests (Spring) related to the features (with a high test coverage).
- [x] I added pre-authorization annotations according to the [guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/#rest-endpoint-best-practices-for-authorization) and checked the course groups for all new REST Calls (security).
- [x] I documented the Java code using JavaDoc style.
#### Client
- [x] **Important**: I implemented the changes with a very good performance, prevented too many (unnecessary) REST calls and made sure the UI is responsive, even with large data.
- [x] I followed the [coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client/).
- [x] Following the [theming guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-design/), I specified colors only in the theming variable files and checked that the changes look consistent in both the light and the dark theme.
- [x] I added multiple integration tests (Jest) related to the features (with a high test coverage), while following the [test guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-tests/).
- [x] I added `authorities` to all new routes and checked the course groups for displaying navigation elements (links, buttons).
- [x] I documented the TypeScript code using JSDoc style.
- [x] I added multiple screenshots/screencasts of my UI changes.
- [x] I translated all newly inserted strings into English and German.

### Motivation and Context

Administrators should be able to request a data export for other users  

### Description

a new POST endpoint in the `DataExportResource` that is only accessible for admins.
Instead of retrieving the currently logged in user from the security context, we fetch the user from the database using the provided login.
Afterward, the flow of events is the same as for a regular data export request and the email and notification is sent to the user the export is requested for.

### Steps for Testing

Prerequisites:
- 1 Administrator
- 1 user with fewer privileges

1. Log in to Artemis as admin
2. Navigate to Privacy Statement
3. Navigate to data export
4. Try to request one for yourself
5. Try to request one for another user
6. In both cases, check that you get the confirmation
7. Log in as a user that is not an admin
8. Do 2-4. 
9. Check that you cannot do 5.

#### Performance Review
- [ ] I (as a reviewer) confirm that the client changes (in particular related to REST calls and UI responsiveness) are implemented with a very good performance 
- [ ] I (as a reviewer) confirm that the server changes (in particular related to database calls) are implemented with a very good performance
#### Code Review
- [x] Code Review 1
- [x] Code Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2

#### Client

| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|---------------------:|
| [data-export-confirmation-dialog.component.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5299/latest/artifact/shared/Coverage-Report-Client-Tests/app/core/legal/data-export/confirmation/data-export-confirmation-dialog.component.ts.html) | 100% | ✅|
| [data-export-confirmation-dialog.service.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5299/latest/artifact/shared/Coverage-Report-Client-Tests/app/core/legal/data-export/confirmation/data-export-confirmation-dialog.service.ts.html) | 100% | ✅ |
| [data-export-request-button.directive.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5299/latest/artifact/shared/Coverage-Report-Client-Tests/app/core/legal/data-export/confirmation/data-export-request-button.directive.ts.html) | 96.42% | ✅ |
| [data-export.component.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5299/latest/artifact/shared/Coverage-Report-Client-Tests/app/core/legal/data-export/data-export.component.ts.html) | 100% | ✅ |
| [data-export.service.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5299/latest/artifact/shared/Coverage-Report-Client-Tests/app/core/legal/data-export/data-export.service.ts.html) | 100% | ✅ |
| [type-ahead-user-search-field.component.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5299/latest/artifact/shared/Coverage-Report-Client-Tests/app/shared/type-ahead-search-field/type-ahead-user-search-field.component.ts.html) | 100% | ✅ |

#### Server

| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|---------------------:|
| [DataExportService.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5299/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.service.dataexport/DataExportService.html) | 99% | ✅ |
| [DataExportResource.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5299/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.web.rest/DataExportResource.html) | 98% | ✅ |
| [AdminDataExportResource.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5299/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.web.rest.admin/AdminDataExportResource.html) | 100% | ✅ |

### Screenshots

https://github.com/ls1intum/Artemis/assets/84102468/689d01c3-e56e-4ec6-ba05-ec46d89be4c9

### CODE REVIEW

**Review by terlan98 — CHANGES_REQUESTED** [review]
Tested on TS6. The English translation keys are missing when the "Request data export for another user" checkbox is enabled:

  Screenshot 2023-08-15 at 15 30 23

File: src/main/java/de/tum/in/www1/artemis/service/dataexport/DataExportService.java [hunk 1]
```diff
@@ -64,6 +63,15 @@ public RequestDataExportDTO requestDataExport() throws IOException {
         return new RequestDataExportDTO(dataExport.getId(), dataExport.getDataExportState(), dataExport.getCreatedDate().atZone(ZoneId.systemDefault()));
     }
 
+    public RequestDataExportDTO requestDataExportForUserAsAdmin(String login) {
```
Comment (MaximilianAnzinger) [inline src/main/java/de/tum/in/www1/artemis/service/dataexport/DataExportService.java:66]: JavaDoc would be nice

File: src/main/java/de/tum/in/www1/artemis/web/rest/admin/AdminDataExportResource.java [hunk 2]
```diff
@@ -0,0 +1,33 @@
+package de.tum.in.www1.artemis.web.rest.admin;
+
+import org.springframework.web.bind.annotation.*;
```
Comment (MaximilianAnzinger) [inline src/main/java/de/tum/in/www1/artemis/web/rest/admin/AdminDataExportResource.java:3]: I think Teamscale is right this time.

File: src/main/webapp/app/shared/type-ahead-search-field/type-ahead-user-search-field.component.html [hunk 3]
```diff
@@ -0,0 +1,33 @@
+<label class="mb-1" for="typeahead-user-login">{{ 'artemisApp.dataExport.loginOrName' | artemisTranslate }}</label>
+<div class="search-container">
+    <input
+        id="typeahead-user-login"
+        type="text"
+        class="form-control"
+        [(ngModel)]="loginOrName"
+        [ngbTypeahead]="search"
+        [placeholder]="'artemisApp.dataExport.loginOrName' | artemisTranslate"
+        (input)="onChange()"
```
Comment (MaximilianAnzinger) [inline src/main/webapp/app/shared/type-ahead-search-field/type-ahead-user-search-field.component.html:10]: is it really necessary to call onChange? I've never seen this in any other component I think.

**Review by MaximilianAnzinger — COMMENTED** [review]
Code looks good. I left some suggestions and a question.

File: src/main/java/de/tum/in/www1/artemis/web/rest/DataExportResource.java [hunk 4]
```diff
@@ -70,6 +71,11 @@ public RequestDataExportDTO requestDataExport() throws IOException {
      */
     private boolean canRequestDataExport() {
         var user = userRepository.getUser();
+        return checkIfUserCanRequest(user);
+
+    }
+
+    private boolean checkIfUserCanRequest(User user) {
```
[the remark this answers is not in the document]
Comment (tobias-lippert) [reply src/main/java/de/tum/in/www1/artemis/web/rest/DataExportResource.java:78]: there's none after I removed the check for the admin endpoint as I think they should always be able to request.
I combined them to one method again.

**Review by Strohgelaender — APPROVED** [review]
lgtm

**Review by Strohgelaender — CHANGES_REQUESTED** [review]
tested in testing session, the client is not using the correct admin endpoint.

**Review by MaximilianAnzinger — APPROVED** [review]
Tested during testing session.

**Review by Strohgelaender — APPROVED** [review]
tested in testing session, works. code lgtm.

**Review by lennart-keller — APPROVED** [review]
Tested during testing session. Works well.

**Review by jakubriegel — APPROVED** [review]
Works (testing session) 👍🏻

**Review by basak-akan — APPROVED** [review]
Tested during the testing session works as described

**Review by RY997 — APPROVED** [review]
Tested in test session

**Review by MaximilianAnzinger — APPROVED** [review]
Tested during testing session.

Thank you for implementing my suggestions / answering my questions. Code LGTM 👍

**Review by nityanandaz — APPROVED** [review]
Code looks good, tested during 2023-08-25 testing session

**Pull request merged**